File: scripts/select-wallpaper.ts
#!/usr/bin/env bun

import { $ } from 'bun';
import { readdir, stat } from 'node:fs/promises';
import { join } from 'node:path';
import { exit } from 'node:process';
import { PATHS } from './utils.ts';

const hyprlockPath = PATHS.HYPR.LOCK;
const hyprpaperPath = PATHS.HYPR.PAPER;
const wallpaperPath = PATHS.WALLPAPERS;
const acceptedExtensions = /\.(jpg|jpeg|png)$/i;

const files = await getAvailableWallpapers();

const selectedFile = (
  await $`printf "%s\n" ${files} | rofi -dmenu -p "wallpapers"`.text()
).trim();

if (!selectedFile) {
  exit(0);
}
const selectedFilePath = join(wallpaperPath, selectedFile);

Promise.all([
  updateHyprpaper(selectedFilePath),
  updateHyprlock(selectedFilePath),
  updateThemes(selectedFilePath),
]);

async function getAvailableWallpapers() {
  const foundPath = await stat(wallpaperPath);
  if (!foundPath.isDirectory()) {
    exit(1);
  }

  const files = (await readdir(wallpaperPath)).filter((file) =>
    acceptedExtensions.test(file),
  );
  if (!files.length) {
    exit(1);
  }

  return files;
}

async function updateHyprpaper(selectedFilePath: string) {
  await $`hyprctl hyprpaper wallpaper ', ${selectedFilePath}'`;

  await updateConfigFile(hyprpaperPath, selectedFilePath, 'wallpaper');
}

async function updateHyprlock(selectedFilePath: string) {
  await updateConfigFile(hyprlockPath, selectedFilePath, 'background');
}

async function updateConfigFile(
  configFilePath: string,
  selectedFilePath: string,
  parentKey: string,
) {
  const lines = (await Bun.file(configFilePath).text()).split('\n');

  const backgroundLine = lines.indexOf(`${parentKey} {`);

  if (backgroundLine === -1) {
    exit(0);
  }

  const pathIndex = findPathKey(backgroundLine, lines);

  lines[pathIndex] = `    path = ${selectedFilePath}`;

  Bun.write(configFilePath, lines.join('\n'));
}

function findPathKey(parentKeyIndex: number, lines: string[]) {
  let pathIndex = -1;
  for (let i = parentKeyIndex + 1; i < lines.length; i++) {
    if (lines[i] === '}') {
      break;
    }

    if (lines[i].includes('path')) {
      pathIndex = i;
      break;
    }
  }

  if (pathIndex === -1) {
    exit(1);
  }

  return pathIndex;
}

async function updateThemes(selectedFilePath: string) {
  await $`matugen --config ${PATHS.MATUGEN} image ${selectedFilePath}`;
}

File: scripts/select-wallpaper.lua
#!/usr/bin/env lua

local script_dir = arg[0]:match("(.*/)") or "./"
package.path = script_dir .. "?.lua;" .. script_dir .. "?/init.lua;" .. package.path

local path = require("lib.path")
local fs = require("lib.fs")

local image_extensions = {".jpg", ".jpeg", ".png"}

local files = fs.list_files(path.WALLPAPERS, image_extensions)

if #files == 0 then
    io.stderr:write("No wallpapers found in " .. path.WALLPAPERS .. "\n")
    os.exit(1)
end

local menu_items = table.concat(files, "\n")

local handle = io.popen('printf "%s" "' .. menu_items .. '" | rofi -dmenu -p "wallpapers"')
if not handle then
    io.stderr:write("Failed to open rofi\n")
    os.exit(1)
end

local selected = handle:read("*a")
handle:close()

selected = selected:gsub("^%s+", ""):gsub("%s+$", "")

if not selected or selected == "" then
    os.exit(0)
end

local selected_path = path.WALLPAPERS .. "/" .. selected

os.execute('hyprctl hyprpaper wallpaper ",' .. selected_path .. '" >/dev/null 2>&1')

local ok, err = fs.update_section_path(path.HYPR_PAPER, "wallpaper", selected_path)
if not ok then
    io.stderr:write("Failed to update hyprpaper.conf: " .. (err or "unknown") .. "\n")
end

ok, err = fs.update_section_path(path.HYPR_LOCK, "background", selected_path)
if not ok then
    io.stderr:write("Failed to update hyprlock.conf: " .. (err or "unknown") .. "\n")
end

os.execute('matugen --config "' ..
    path.MATUGEN .. '" --source-color-index 0 image "' .. selected_path .. '" >/dev/null 2>&1')
